io_scene_vrm/importer/vrm_animation_importer.py

```diff
@@ -0,0 +1,856 @@
+        timestamp = (
+            frame_count * context.scene.render.fps_base / context.scene.render.fps
+        )
+        assign_humanoid_keyframe(
+            armature,
+            node_rest_pose_tree,
+            node_index_to_human_bone_name,
+            node_index_to_translation_keyframes,
+            node_index_to_rotation_keyframes,
+            frame_count,
+            timestamp,
+            humanoid_parent_rest_world_matrix=Matrix(),
+            intermediate_rest_local_matrix=Matrix(),
+            intermediate_pose_local_matrix=Matrix(),
+        )
+        assign_expression_keyframe(
+            armature_data,
+            expression_name_to_default_preview_value,
+            expression_name_to_translation_keyframes,
+            frame_count,
+            timestamp,
+        )
+
+    return {"FINISHED"}
+
+
+def dump(v: Union[Matrix, Vector, Quaternion, float, int]) -> str:
+    if logger.level > logging.DEBUG:
+        return "(omit)"
+
+    if isinstance(v, (float, int)):
+        return str(v)
+
+    if isinstance(v, Matrix):
+        t, r, s = v.decompose()
+        return f"Matrix(T={dump(t)},R={dump(r)},S={dump(s)})"
+
+    if isinstance(v, Vector):
+        return f"({v.x:.3f},{v.y:.3f},{v.z:.3f})"
+
+    x, y, z = [round(math.degrees(xyz)) for xyz in v.to_euler("XYZ")[:]]
+    return f"Euler({x},{y},{z})"
+
+
+def assign_expression_keyframe(
+    armature_data: bpy.types.Armature,
+    expression_name_to_default_preview_value: dict[str, float],
+    expression_name_to_translation_keyframes: dict[
+        str, tuple[tuple[float, Vector], ...]
+    ],
+    frame_count: int,
+    timestamp: float,
+) -> None:
+    expressions = armature_data.vrm_addon_extension.vrm1.expressions
+    if not isinstance(expressions, Vrm1ExpressionsPropertyGroup):
+        # make static type checker happy
+        raise AssertionError(f"{expressions=} is not a Vrm1ExpressionsPropertyGroup")
+    expression_name_to_expression = expressions.all_name_to_expression_dict()
+    for (
+        expression_name,
+        translation_keyframes,
+    ) in expression_name_to_translation_keyframes.items():
+        expression = expression_name_to_expression.get(expression_name)
+        if not expression:
+            continue
+
+        if translation_keyframes:
+            _, preview = translation_keyframes[0][1].x
+            for keyframe_timestamp, keyframe_translation in translation_keyframes:
+                if keyframe_timestamp >= timestamp:
+                    preview = keyframe_translation.x
+                    break
+        else:
+            preview = (
+                expression_name_to_default_preview_value.get(expression_name) or 0.0
+            )
+
+        current_preview = expression.preview
+        expression.preview = preview
+        expression.keyframe_insert(data_path="preview", frame=frame_count)
+        expression.preview = current_preview
+
+
+def assign_humanoid_keyframe(
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This function contains 71 lines of code, not including blank lines or lines with only comments, Python punctuation characters, identifiers, or literals. By comparison, 99% of the functions in the CodeGuru reference dataset contain fewer lines of code. Large functions might be difficult to read and have logic that is hard to understand and test. We recommend that you simplify this function or break it into multiple functions.